src/rust/src/buf.rs

```diff
@@ -46,47 +40,130 @@ fn _extract_buffer_length<'p>(
     Ok((bufobj, ptrval))
 }
 
-impl<'a> CffiBuf<'a> {
-    pub(crate) fn from_bytes(py: pyo3::Python<'a>, buf: &'a [u8]) -> Self {
-        CffiBuf {
-            pyobj: py.None().into_bound(py),
-            _bufobj: py.None().into_bound(py),
-            buf,
-        }
+#[cfg(Py_3_11)]
+mod pybuffer_impl {
+    use pyo3::buffer::PyBuffer;
+    use pyo3::types::{PyAnyMethods, PyBytes};
+    use std::slice;
+
+    pub(crate) struct CffiBuf<'p> {
+        pyobj: pyo3::Bound<'p, pyo3::PyAny>,
+        _bufobj: PyBuffer<u8>,
+        buf: &'p [u8],
     }
 
-    pub(crate) fn as_bytes(&self) -> &[u8] {
-        self.buf
+    impl<'a> CffiBuf<'a> {
+        pub(crate) fn from_bytes(py: pyo3::Python<'a>, buf: &'a [u8]) -> Self {
+            let py_bytes = PyBytes::new(py, buf);
+            let pybuffer = PyBuffer::<u8>::get(&py_bytes)
+                .expect("Cannot convert \"bytes\" instance to a buffer.");
+            CffiBuf {
+                pyobj: py.None().into_bound(py),
+                _bufobj: pybuffer,
+                buf,
+            }
+        }
+
+        pub(crate) fn as_bytes(&self) -> &[u8] {
+            self.buf
+        }
+
+        pub(crate) fn into_pyobj(self) -> pyo3::Bound<'a, pyo3::PyAny> {
+            self.pyobj
+        }
     }
 
-    pub(crate) fn into_pyobj(self) -> pyo3::Bound<'a, pyo3::PyAny> {
-        self.pyobj
+    impl<'a> pyo3::conversion::FromPyObject<'a> for CffiBuf<'a> {
+        fn extract_bound(pyobj: &pyo3::Bound<'a, pyo3::PyAny>) -> pyo3::PyResult<Self> {
+            let bufobj = PyBuffer::<u8>::get(pyobj).map_err(|_| {
+                let errmsg = if pyobj.is_instance_of::<pyo3::types::PyString>() {
+                    format!(
+                        "Cannot convert \"{}\" instance to a buffer.\nDid you mean to pass a bytestring instead?",
+                        pyobj.get_type()
+                    )
+                } else {
+                    format!("Cannot convert \"{}\" instance to a buffer.", pyobj.get_type())
+                };
+                pyo3::exceptions::PyTypeError::new_err(errmsg)
+            })?;
+            let len = pyobj.len()?;
+            let buf = if len == 0 {
+                &[]
+            } else {
+                // SAFETY: pyo3::buffer::PyBuffer ensures that we have a valid ptr
```

Comment:
Probably much safer now since we know its not released until `drop`
https://github.com/PyO3/pyo3/blob/c25c6f773dd023720b4e6b3969d5e99fef57433d/src/buffer.rs#L198
https://github.com/PyO3/pyo3/blob/c25c6f773dd023720b4e6b3969d5e99fef57433d/src/buffer.rs#L613